// cypress-ntlm-auth/dist/proxy/config.server.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const getPort = require('get-port');
const inversify_1 = require("inversify");
const dependency_injection_types_1 = require("./dependency.injection.types");
let ConfigServer = class ConfigServer {
    constructor(expressServer, configController, debug) {
        this._configApiUrl = undefined;
        this.initDone = false;
        this._expressServer = expressServer;
        this._configController = configController;
        this._debug = debug;
    }
    get configApiUrl() {
        if (this._configApiUrl) {
            return this._configApiUrl;
        }
        throw new Error('Cannot get configApiUrl, ConfigServer not started!');
    }
    init() {
        if (this.initDone) {
            return;
        }
        this._expressServer.use('/', this._configController.router);
        this.initDone = true;
    }
    start(port) {
        return __awaiter(this, void 0, void 0, function* () {
            this.init();
            try {
                if (!port) {
                    port = yield getPort();
                    if (port === undefined) {
                        this._debug.log('Cannot find free port');
                        throw new Error('Cannot find free port');
                    }
                }
                this._configApiUrl = yield this._expressServer.listen(port);
                this._debug.log('NTLM auth config API listening on port:', port);
                return this._configApiUrl;
            }
            catch (err) {
                this._debug.log('Cannot start NTLM auth config API');
                throw err;
            }
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            this._debug.log('Shutting down config API');
            try {
                yield this._expressServer.close();
                this._configApiUrl = undefined;
                this._debug.log('NTLM auth config API stopped');
            }
            catch (err) {
                this._debug.log('Cannot stop NTLM auth config API');
                throw err;
            }
        });
    }
};
ConfigServer = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(dependency_injection_types_1.TYPES.IExpressServerFacade)),
    __param(1, inversify_1.inject(dependency_injection_types_1.TYPES.IConfigController)),
    __param(2, inversify_1.inject(dependency_injection_types_1.TYPES.IDebugLogger)),
    __metadata("design:paramtypes", [Object, Object, Object])
], ConfigServer);
exports.ConfigServer = ConfigServer;
