// cypress-ntlm-auth/dist/proxy/config.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = require("express");
const config_validator_1 = require("../util/config.validator");
const url_converter_1 = require("../util/url.converter");
const inversify_1 = require("inversify");
const events_1 = require("events");
const dependency_injection_types_1 = require("./dependency.injection.types");
const sso_config_validator_1 = require("../util/sso.config.validator");
let ConfigController = class ConfigController {
    constructor(configStore, debug) {
        this.router = express_1.Router();
        this.configApiEvent = new events_1.EventEmitter();
        this._configStore = configStore;
        this._debug = debug;
        this.router.post('/ntlm-config', (req, res) => this.ntlmConfig(req, res));
        this.router.post('/ntlm-sso', (req, res) => this.ntlmSso(req, res));
        this.router.post('/reset', (req, res) => this.reset(req, res));
        this.router.get('/alive', (req, res) => this.alive(req, res));
        this.router.post('/quit', (req, res) => this.quit(req, res));
    }
    ntlmConfig(req, res) {
        let validateResult = config_validator_1.ConfigValidator.validate(req.body);
        if (!validateResult.ok) {
            res.status(400).send('Config parse error. ' + validateResult.message);
        }
        else {
            this._debug.log('Received valid NTLM config update');
            let config = req.body;
            let ntlmHostUrl = url_converter_1.toCompleteUrl(config.ntlmHost, false);
            if (this._configStore.exists(ntlmHostUrl)) {
                // Trigger removal of existing authentication cache
                this.configApiEvent.emit('configUpdate', ntlmHostUrl);
                this._debug.log('Updating host', ntlmHostUrl.href);
            }
            else {
                this._debug.log('Added new host', ntlmHostUrl.href);
            }
            this._configStore.updateConfig(config);
            res.sendStatus(200);
        }
    }
    ntlmSso(req, res) {
        let validateResult = sso_config_validator_1.SsoConfigValidator.validate(req.body);
        if (!validateResult.ok) {
            res.status(400).send('SSO config parse error. ' + validateResult.message);
        }
        else {
            this._debug.log('Received valid NTLM SSO config');
            let config = req.body;
            this._configStore.setSsoConfig(config);
            res.sendStatus(200);
        }
    }
    reset(req, res) {
        this._debug.log('Received reset');
        this.configApiEvent.emit('reset');
        res.sendStatus(200);
    }
    alive(req, res) {
        this._debug.log('Received alive');
        res.sendStatus(200);
    }
    quit(req, res) {
        this._debug.log('Received quit');
        res.status(200).send('Over and out!');
        let keepPortsFile = req.body && req.body.keepPortsFile;
        this.configApiEvent.emit('quit', keepPortsFile);
    }
};
ConfigController = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(dependency_injection_types_1.TYPES.IConfigStore)),
    __param(1, inversify_1.inject(dependency_injection_types_1.TYPES.IDebugLogger)),
    __metadata("design:paramtypes", [Object, Object])
], ConfigController);
exports.ConfigController = ConfigController;
