// cypress-ntlm-auth/dist/proxy/connection.context.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ntlm_state_enum_1 = require("../models/ntlm.state.enum");
const inversify_1 = require("inversify");
let ConnectionContext = class ConnectionContext {
    constructor() {
        this._ntlmState = ntlm_state_enum_1.NtlmStateEnum.NotAuthenticated;
        this._requestBody = Buffer.alloc(0);
        this._clientAddress = "";
    }
    get agent() {
        return this._agent;
    }
    set agent(agent) {
        this._agent = agent;
    }
    get winSso() {
        if (!this._winSso) {
            throw new Error("WinSso not initialized for context");
        }
        return this._winSso;
    }
    set winSso(winSso) {
        this._winSso = winSso;
    }
    get peerCert() {
        return this._peerCert;
    }
    set peerCert(peerCert) {
        this._peerCert = peerCert;
    }
    get clientAddress() {
        return this._clientAddress;
    }
    set clientAddress(clientAddress) {
        this._clientAddress = clientAddress;
    }
    isNewOrAuthenticated(ntlmHostUrl) {
        let auth = this._ntlmHost === undefined ||
            (this._ntlmHost !== undefined &&
                this._ntlmHost.href === ntlmHostUrl.href &&
                this._ntlmState === ntlm_state_enum_1.NtlmStateEnum.Authenticated);
        return auth;
    }
    matchHostOrNew(ntlmHostUrl) {
        return (this._ntlmHost === undefined || this._ntlmHost.href === ntlmHostUrl.href);
    }
    getState(ntlmHostUrl) {
        if (this._ntlmHost && ntlmHostUrl.href === this._ntlmHost.href) {
            return this._ntlmState;
        }
        return ntlm_state_enum_1.NtlmStateEnum.NotAuthenticated;
    }
    setState(ntlmHostUrl, authState) {
        this._ntlmHost = ntlmHostUrl;
        this._ntlmState = authState;
    }
    clearRequestBody() {
        this._requestBody = Buffer.alloc(0);
    }
    addToRequestBody(chunk) {
        this._requestBody = Buffer.concat([this._requestBody, chunk]);
    }
    getRequestBody() {
        return this._requestBody;
    }
    destroy() {
        if (this._agent.destroy) {
            this._agent.destroy(); // Destroys any sockets to servers
        }
        if (this._winSso) {
            this._winSso.freeAuthContext();
        }
    }
};
ConnectionContext = __decorate([
    inversify_1.injectable()
], ConnectionContext);
exports.ConnectionContext = ConnectionContext;
